scipy/stats/_odds_ratio.py

```diff
@@ -0,0 +1,479 @@
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from scipy.special import ndtr, ndtri
+from scipy.optimize import brentq
+from .distributions import nchypergeom_fisher, hypergeom
+from .stats import fisher_exact
+
+
+def _sample_odds_ratio(table):
+    """
+    Given a table [[a, b], [c, d]], compute a*d/(b*c).
+
+    Return nan if the numerator and denominator are 0.
+    Return inf if just the denominator is 0.
+    """
+    # table must be a 2x2 numpy array.
+    if table[1, 0] > 0 and table[0, 1] > 0:
+        oddsratio = table[0, 0] * table[1, 1] / (table[1, 0] * table[0, 1])
+    elif table[0, 0] == 0 or table[1, 1] == 0:
+        oddsratio = np.nan
+    else:
+        oddsratio = np.inf
+    return oddsratio
+
+
+def _solve(func):
+    """
+    Solve func(nc) = 0.  func must be an increasing function.
+    """
+    # We could just as well call the variable `x` instead of `nc`, but we
+    # always call this function with functions for which nc (the noncentrality
+    # parameter) is the variable for which we are solving.
+    nc = 1.0
+    value = func(nc)
+    if value == 0:
+        return nc
+
+    # Multiplicative factor by which to increase or decrease nc when
+    # searching for a bracketing interval.
+    factor = 25.0
+    # Find a bracketing interval.
+    if value > 0:
+        nc /= factor
+        while func(nc) > 0:
+            nc /= factor
+        lo = nc
+        hi = factor*nc
+    else:
+        nc *= factor
+        while func(nc) < 0:
+            nc *= factor
+        lo = nc/factor
+        hi = nc
+
+    # lo and hi bracket the solution for nc.
+    nc = brentq(func, lo, hi, xtol=1e-13)
+    return nc
+
+
+def _nc_hypergeom_mean_inverse(x, total, ngood, nsample):
+    """
+    For the given total, ngood, and nsample, find the noncentrality
+    parameter of Fisher's noncentral hypergeometric distribution whose
+    mean is x.
+    """
+    nc = _solve(lambda nc: nchypergeom_fisher.mean(total, ngood, nsample,
+                                                   nc) - x)
```

Comment:
Converting between different conventions is annoying, so I haven't. Assuming the Breslow and Day reference equation 2 is an equivalent formulation of Fisher's NCHG distribution with these arguments, this looks right.

Do you want me to check it carefully, or shall I assume that you did and that it is unlikely that the numbers this function produces would match R otherwise?

Reply:
This is probably not quite what you are looking for, but I updated the code to use parameter names that are consistent with the names used in `stats.hypergeom` and `stats.nchypergeom_fisher`.

For a contingency table [[a, b], [c, d]], Breslow and Day's notation is defined in their "equation" (4.1).  The correspondence to the SciPy parameters is
```
SciPy    Breslow & Day     Meaning in terms of 2x2 contingency table
  k           a                    table[0, 0]
  M           N                    sum(table)
  n           n1                   table[0].sum() 
  N           m1                   table[: 0].sum()     
```
In Breslow and Day's notation, pr(a | n1, n0, m1, m0, Ψ) is the PMF for the noncentral hypergeometric distribution; that parameterization is redundant, since n0 + n1 = m0 + m1.